### PR TITLE
Refactor special track elements list part 2

### LINKED ISSUE #19964: A 2nd S-bend?
### Operating System

Windows 11, 64-Bits

### OpenRCT2 build

OpenRCT2, v0.4.4-147-ge88d606 (e88d606 on develop) provided by GitHub

### Base game

RollerCoaster Tycoon 2

### Area(s) with this issue?

This bug is a graphical glitch or error

### Describe the issue

there is the standard S-bends, as usual, but I stumple on a glitched version of them, the video explain and illustrates the issue in real time, as at first, I though it was because of weird placement, but it seem there is just in fact, a 2nd S-bend existing?!?

### Steps to reproduce

1. Go to any parks
2. Build any rides whit S-bends as usual E. G. single rail coaster
3. Go to the special list
4. find 2 S-bends, which the 2nd time occurence is a glitched version

### Attachments

https://drive.google.com/file/d/1t-k709Z7p5Tb4tvPfoaVo6qlxVR13lUR/view?usp=sharing

### PR DESCRIPTION
**Pull request by spacek531**
Part 2 of https://github.com/OpenRCT2/OpenRCT2/pull/19922
Fixes #19964

There are a lot of files changed, so this PR is structured to be easy to follow by looking at individual commits.

This PR refactors how the special track element list is created. The refactors include:
- Divorces creating the list from RideConstructionWindow for modularity, including potential plugin API
- Adds list of special track elements which can be re-ordered for more sensible organization
- Same-named track elements will collapse into as few entries as possible
- Separates up helixes from down helixes in track flags

Track flag separation was done to remove special code related to alpine coaster. Refactors to ride construction window were tested against and confirmed working with:
- alpine coaster
- suspended swinging coaster
- hybrid coaster
- inverted coaster

~~Currently draft as I am waiting for community feedback on potential alternate orderings.~~ I have dropped the reordering part of this PR, any reordering can be done in a later PR.

### CODE REVIEW

File: src/openrct2-ui/windows/RideConstruction.cpp [hunk 1]
```diff
@@ -26,6 +28,7 @@
 #include <openrct2/actions/TrackSetBrakeSpeedAction.h>
 #include <openrct2/audio/audio.h>
 #include <openrct2/config/Config.h>
+#include <openrct2/core/BitSet.hpp>
```
Comment (spacek531) [inline src/openrct2-ui/windows/RideConstruction.cpp:31]: todo: remove unnecessary include

File: src/openrct2/ride/TrackData.cpp [hunk 2]
```diff
@@ -6365,29 +6365,29 @@ static constexpr TrackDefinition TrackDefinitions[] =
...
+    { TRACK_HELIX_UP_BANKED_QUARTER,     TRACK_SLOPE_NONE,      TRACK_SLOPE_NONE,           TRACK_BANK_LEFT,        TRACK_BANK_LEFT,        0                  },  // ELEM_LEFT_QUARTER_BANKED_HELIX_LARGE_UP
+    { TRACK_HELIX_UP_BANKED_QUARTER,     TRACK_SLOPE_NONE,      TRACK_SLOPE_NONE,           TRACK_BANK_RIGHT,       TRACK_BANK_RIGHT,       0                  },  // ELEM_RIGHT_QUARTER_BANKED_HELIX_LARGE_UP
+    { TRACK_HELIX_DOWN_BANKED_QUARTER,   TRACK_SLOPE_NONE,      TRACK_SLOPE_NONE,           TRACK_BANK_LEFT,        TRACK_BANK_LEFT,        0                  },  // ELEM_LEFT_QUARTER_BANKED_HELIX_LARGE_DOWN
+    { TRACK_HELIX_DOWN_BANKED_QUARTER,   TRACK_SLOPE_NONE,      TRACK_SLOPE_NONE,           TRACK_BANK_RIGHT,       TRACK_BANK_RIGHT,       0                  },  // ELEM_RIGHT_QUARTER_BANKED_HELIX_LARGE_DOWN
+    { TRACK_HELIX_DOWN_UNBANKED_QUARTER, TRACK_SLOPE_NONE,      TRACK_SLOPE_NONE,           TRACK_BANK_NONE,        TRACK_BANK_NONE,        0                  },  // ELEM_LEFT_QUARTER_HELIX_LARGE_UP
```
Comment (spacek531) [inline src/openrct2/ride/TrackData.cpp:6387]: TRACK_HELIX_UP_UNBANKED_QUARTER

File: src/openrct2-ui/windows/RideConstruction.cpp [hunk 3]
```diff
@@ -199,10 +201,7 @@ class RideConstructionWindow final : public Window
 {
 private:
     uint8_t _currentlyShowingBrakeOrBoosterSpeed{};
-    uint64_t _currentDisabledSpecialTrackPieces{};
-    track_type_t _currentPossibleRideConfigurations[64]{};
-    uint16_t _numCurrentPossibleRideConfigurations{};
-    uint16_t _numCurrentPossibleSpecialTrackPieces{};
+    SpecialElementsList SpecialElementList;
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/RideConstruction.cpp:204]: this should still have a _ prefix as its a private member variable. `SpecialElementsList _specialElementsList`. List implies linked list I would avoid that term for this. Perhaps work in Array or tbh its probably pretty obvious if its plural `_specialElements`.

File: src/openrct2-ui/ride/Construction.h [hunk 4]
```diff
@@ -10,5 +10,153 @@
...
+
+struct SpecialElement
+{
+    track_type_t TrackType;
+    bool Disabled;
+};
+
+struct SpecialElementsList
+{
+    int32_t NumElements{};
```
Comment (duncanspumpkin) [inline src/openrct2-ui/ride/Construction.h:154]: If you need a NumElements i'd prefer you switched Elements to being a std::vector and you grow to NumElements. That makes things a bit more sensible.

File: src/openrct2-ui/windows/RideConstruction.cpp (hunk 3 above)
Comment (spacek531) [reply src/openrct2-ui/windows/RideConstruction.cpp:204]: I will call it SpecialElementsDropdownState

**Comment by spacek531** [comment]
I have simplified the collating logic to only look at the last element in the list. When the list is open, it is being built every frame, and I decided to indulge in this micro-optimization. This does mean that all track elements with the same name must be sequential, so I reordered the elements the minimum amount to accomplish this. The only differences will be that all the multidim quarter loops will appear above block brake in the list (previously, 3 of the 4 appeared below), and the OpenRCT2-only flying half loops will appear above block OpenRCT2-only flying large half loops, instead of below.

**Comment by spacek531** [comment]
Fixes https://github.com/OpenRCT2/OpenRCT2/issues/19964

**Comment by spacek531** [comment]
Any chance this can make it into 0.4.5? It fixes a bug I introduced in part 1.

File: src/openrct2-ui/ride/Construction.cpp [hunk 5]
```diff
@@ -36,3 +40,60 @@ void RideConstructNew(RideSelection listItem)
...
+                || TrackPieceDirectionIsDiagonal(ted.Coordinates.rotation_begin) != buildDirectionIsDiagonal;
+        }
+
+        // Additional tower bases can only be built if the ride allows for it (elevator)
+        if (trackType == TrackElemType::TowerBase
+            && !currentRide.GetRideTypeDescriptor().HasFlag(RIDE_TYPE_FLAG_ALLOW_EXTRA_TOWER_BASES))
+            entryIsDisabled = true;
+
+        // Check if a previous element exists, to collate entries if possible
+        if (list.Elements.size() > 0
```
Comment (ZehMatt) [inline src/openrct2-ui/ride/Construction.cpp:79]:
```suggestion
        if (!list.Elements.empty()
```

File: src/openrct2-ui/ride/Construction.cpp [hunk 6]
```diff
@@ -36,3 +40,60 @@ void RideConstructNew(RideSelection listItem)
...
+            && GetTrackElementDescriptor(list.Elements.back().TrackType).Description == ted.Description)
+        {
+            // If the current element is disabled, do not add current element.
+            if (entryIsDisabled)
+                continue;
+            // If the previous element is disabled and current element is enabled, replace the previous element
+            if (list.Elements.back().Disabled && !entryIsDisabled)
```
Comment (ZehMatt) [inline src/openrct2-ui/ride/Construction.cpp:86]:
```suggestion
            auto& lastElement = list.Elements.back();
            if (lastElement.Disabled && !entryIsDisabled)
```
the lines below should use the variable as well.

**Review by ZehMatt — CHANGES_REQUESTED** [review]
I would prefer the fix is separated from this PR, also left a few comments otherwise fine by me.

**Review by ZehMatt — APPROVED** [review]
I've tested this and it all seems ok, built a few rides with and checked the special elements.

**Pull request merged**